File: R/metrics.R
#' The true skill statistic
#'
#' Sensitivity plus specificity minus one, at the threshold that maximises it. This is the metric
#' species distribution modelling reports, and the one the shipped presence-absence response is
#' scored by.
#'
#' A cut may only fall between distinct predictions: units sharing a prediction are decided
#' together, so the same score comes back whatever order they arrived in. A cell holding only
#' presences or only absences has no skill to measure and returns `NA` rather than a number.
#'
#' The threshold is chosen on the same units the score is then read on, which is how the metric is
#' defined in the literature and how it is defined here, and it inflates the level where presences
#' are thin. [tss_inflation()] measures that inflation for a given design, and it cancels in the
#' paired differences [paired_contrast()] takes. Given a `threshold` learned elsewhere, the score
#' is read at that cut instead, which is what [select_grain()] reports under `threshold =` with a
#' cut learned on the inner folds.
#'
#' @param y Observed presence-absence, `0`/`1` or logical.
#' @param p Predicted scores for the same units, in the same order. Higher means presence.
#' @param threshold `NULL` for the maximum over every cut, or one cut, presence being predicted at
#'   `p >= threshold`.
#'
#' @return One number, or `NA` where the cell defines none.
#'
#' @examples
#' tss(c(0, 0, 1, 1), c(0.1, 0.2, 0.8, 0.9))
#' tss(c(0, 0, 1, 1), c(0.9, 0.8, 0.2, 0.1))
#' tss(c(0, 0, 0, 0), c(0.1, 0.2, 0.8, 0.9))
#' tss(c(0, 0, 1, 1), c(0.1, 0.6, 0.8, 0.9), threshold = 0.5)
#'
#' @export
tss <- function(y, p, threshold = NULL) {
  if (!is.null(threshold)) {
    y <- .check_labels(y, p)
    if (is.null(y) || length(threshold) != 1L || !is.finite(threshold)) {
      return(NA_real_)
    }
    hit <- as.numeric(p) >= threshold
    return(mean(hit[y == 1L]) - mean(hit[y == 0L]))
  }
  s <- .sweep(y, p)
  if (is.null(s)) {
    return(NA_real_)
  }
  max(s$tp / s$n_pos - s$fp / s$n_neg)
}

#' The area under the ROC curve
#'
#' The whole curve rather than its best point. TSS is a maximum over cuts, so a small change in a
#' prediction often moves it not at all and then moves it a long way; the area responds to every
#' reordering, which is what makes it the steadier reading when many rescorings are compared, as in
#' [occlusion()]. Tied predictions take the average rank.
#'
#' @inheritParams tss
#'
#' @return One number, or `NA` where the cell defines none.
#'
#' @examples
#' roc_auc(c(0, 0, 1, 1), c(0.1, 0.2, 0.8, 0.9))
#'
#' @export
roc_auc <- function(y, p) {
  y <- .check_labels(y, p)
  if (is.null(y)) {
    return(NA_real_)
  }
  n_pos <- sum(y == 1L)
  n_neg <- length(y) - n_pos
  (sum(rank(p)[y == 1L]) - n_pos * (n_pos + 1) / 2) / (n_pos * n_neg)
}

#' Cohen's kappa, and where two models disagree
#'
#' A chance-corrected agreement rate on a two-by-two table. Read against the observed response it
#' is a skill score beside [tss()]; read between two models' decisions on the same units it says
#' where the two part company.
#'
#' Kappa is read at a threshold rather than maximised over one, so the rule that picks the
#' threshold is part of the statistic. `"youden"` is the operating point [tss()] is defined at and
#' inherits its selection bias; `"kappa"` maximises kappa itself and inherits the analogous bias;
#' `"prevalence"` cuts at the observed presence rate, which selects nothing from the labels and is
#' the rule to read an absolute level at.
#'
#' @inheritParams tss
#' @param rule Threshold rule: `"youden"`, `"kappa"` or `"prevalence"`.
#'
#' @return For `kappa_score()`, one number. For `decision_threshold()`, the cut itself, applied as
#'   `p >= threshold`. For `model_agreement()`, a one-row data frame carrying the agreement kappa
#'   between two models cut by the same rule, the share of units they decide differently, and how
#'   often each is the one that is right there.
#'
#' @examples
#' y <- c(0, 0, 0, 1, 1, 1)
#' kappa_score(y, c(0.1, 0.2, 0.6, 0.4, 0.8, 0.9))
#' model_agreement(y, c(0.1, 0.2, 0.6, 0.4, 0.8, 0.9), c(0.2, 0.1, 0.3, 0.7, 0.9, 0.8))
#'
#' @export
kappa_score <- function(y, p, rule = c("youden", "kappa", "prevalence")) {
  thr <- decision_threshold(y, p, rule)
  if (!is.finite(thr)) {
    return(NA_real_)
  }
  .kappa_table(.labels(y), as.integer(p >= thr))
}

#' @rdname kappa_score
#' @export
decision_threshold <- function(y, p, rule = c("youden", "kappa", "prevalence")) {
  rule <- match.arg(rule)
  s <- .sweep(y, p)
  if (is.null(s)) {
    return(NA_real_)
  }
  if (rule == "prevalence") {
    return(unname(stats::quantile(as.numeric(p), 1 - s$n_pos / (s$n_pos + s$n_neg), type = 7)))
  }
  if (rule == "youden") {
    return(s$thr[which.max(s$tp / s$n_pos - s$fp / s$n_neg)])
  }
  n <- s$n_pos + s$n_neg
  fn <- s$n_pos - s$tp
  tn <- s$n_neg - s$fp
  po <- (s$tp + tn) / n
  pe <- ((s$tp + s$fp) * s$n_pos + (fn + tn) * s$n_neg) / n^2
  k <- ifelse(pe >= 1, -Inf, (po - pe) / (1 - pe))
  s$thr[which.max(k)]
}

#' @rdname kappa_score
#' @param p_a,p_b Two models' predictions for the same units.
#' @export
model_agreement <- function(y, p_a, p_b, rule = c("youden", "kappa", "prevalence")) {
  rule <- match.arg(rule)
  y <- .labels(y)
  ta <- decision_threshold(y, p_a, rule)
  tb <- decision_threshold(y, p_b, rule)
  if (!is.finite(ta) || !is.finite(tb)) {
    return(data.frame(kappa = NA_real_, n = length(y), n_disagree = NA_integer_,
                      share_disagree = NA_real_, a_right = NA_integer_, b_right = NA_integer_))
  }
  da <- as.integer(p_a >= ta)
  db <- as.integer(p_b >= tb)
  diff <- da != db
  data.frame(kappa = .kappa_table(da, db), n = length(y), n_disagree = sum(diff),
             share_disagree = mean(diff),
             a_right = sum(diff & da == y), b_right = sum(diff & db == y))
}

.kappa_table <- function(a, b) {
  n <- length(a)
  both <- sum(a == 1L & b == 1L)
  neither <- sum(a == 0L & b == 0L)
  a_only <- sum(a == 1L & b == 0L)
  b_only <- sum(a == 0L & b == 1L)
  po <- (both + neither) / n
  pe <- ((both + a_only) * (both + b_only) + (b_only + neither) * (a_only + neither)) / n^2
  if (pe >= 1) NA_real_ else (po - pe) / (1 - pe)
}

# The response as 0/1 integers, refused where it is not one: checked on the values as given,
# because coercing first would read 0.6 as 0 and pass a response that was never binary.
.labels <- function(y) {
  if (!all(y %in% c(0, 1))) {
    stop("`y` must be presence-absence, 0/1 or logical.", call. = FALSE)
  }
  as.integer(y)
}

.check_labels <- function(y, p) {
  y <- .labels(y)
  if (length(y) != length(p)) {
    stop("`y` and `p` must be the same length, got ", length(y), " and ", length(p), ".",
         call. = FALSE)
  }
  if (anyNA(p) || any(!is.finite(as.numeric(p)))) {
    return(NULL)
  }
  n_pos <- sum(y == 1L)
  if (n_pos == 0L || n_pos == length(y)) NULL else y
}

# The one place a score becomes a set of cuts. Ties are decided together, so a cut can only fall
# between distinct predictions; both languages read every threshold metric off this.
.sweep <- function(y, p) {
  y <- .check_labels(y, p)
  if (is.null(y)) {
    return(NULL)
  }
  p <- as.numeric(p)
  o <- order(p, decreasing = TRUE, method = "radix")
  ys <- y[o]
  ps <- p[o]
  keep <- c(ps[-length(ps)] != ps[-1L], TRUE)
  list(thr = ps[keep], tp = cumsum(ys)[keep], fp = cumsum(1L - ys)[keep],
       n_pos = sum(y == 1L), n_neg = sum(y == 0L))
}

File: python/timesift/metrics.py
"""Threshold metrics on held-out predictions.

A cut may only fall between distinct predictions: units sharing a prediction are decided together,
so the same score comes back whatever order they arrived in and both language sides read every
threshold metric off the same rule.
"""

from __future__ import annotations

import numpy as np

THRESHOLD_RULES = ("youden", "kappa", "prevalence")


def _labels(y) -> np.ndarray:
    """The response as 0/1 integers, refused where it is not one: checked on the values as given,
    because coercing first would read 0.6 as 0 and pass a response that was never binary."""
    y = np.asarray(y, dtype=np.float64)
    if not np.isin(y, (0.0, 1.0)).all():
        raise ValueError("`y` must be presence-absence, 0/1 or logical.")
    return y.astype(np.int64)


def _sweep(y, p):
    y = _labels(y)
    p = np.asarray(p, dtype=np.float64)
    if y.shape != p.shape:
        raise ValueError("`y` and `p` must be the same length")
    if not np.isfinite(p).all():
        return None
    n_pos, n_neg = int(y.sum()), int((1 - y).sum())
    if n_pos == 0 or n_neg == 0:
        return None
    order = np.argsort(-p, kind="stable")
    ys, ps = y[order], p[order]
    keep = np.empty(len(ps), dtype=bool)
    keep[:-1] = ps[:-1] != ps[1:]
    keep[-1] = True
    return dict(thr=ps[keep], tp=np.cumsum(ys)[keep].astype(float),
                fp=np.cumsum(1 - ys)[keep].astype(float), n_pos=n_pos, n_neg=n_neg)


def tss(y, p, threshold=None) -> float:
    """Sensitivity plus specificity minus one, at the threshold that maximises it.

    Given a ``threshold`` learned elsewhere, the score is read at that cut instead, presence being
    predicted at ``p >= threshold``.
    """
    if threshold is not None:
        s = _sweep(y, p)
        if s is None or not np.isfinite(threshold):
            return float("nan")
        y = _labels(y)
        hit = np.asarray(p, dtype=np.float64) >= threshold
        return float(hit[y == 1].mean() - hit[y == 0].mean())
    s = _sweep(y, p)
    if s is None:
        return float("nan")
    return float(np.max(s["tp"] / s["n_pos"] - s["fp"] / s["n_neg"]))


def roc_auc(y, p) -> float:
    """The area under the ROC curve, as the rank sum of the presences. Ties take the average rank."""
    y = _labels(y)
    p = np.asarray(p, dtype=np.float64)
    n_pos, n_neg = int(y.sum()), int((1 - y).sum())
    if n_pos == 0 or n_neg == 0 or not np.isfinite(p).all():
        return float("nan")
    order = np.argsort(p, kind="stable")
    ranks = np.empty(len(p), dtype=np.float64)
    ranks[order] = np.arange(1, len(p) + 1)
    # average rank within each run of tied predictions
    ps = p[order]
    start = 0
    for i in range(1, len(ps) + 1):
        if i == len(ps) or ps[i] != ps[start]:
            ranks[order[start:i]] = (start + i + 1) / 2
            start = i
    return float((ranks[y == 1].sum() - n_pos * (n_pos + 1) / 2) / (n_pos * n_neg))


def decision_threshold(y, p, rule: str = "youden") -> float:
    """The probability cut a rule selects. Presence is predicted at ``p >= threshold``."""
    if rule not in THRESHOLD_RULES:
        raise ValueError(f"rule must be one of {THRESHOLD_RULES}, got {rule!r}")
    s = _sweep(y, p)
    if s is None:
        return float("nan")
    if rule == "prevalence":
        return float(np.quantile(np.asarray(p, dtype=np.float64),
                                 1 - s["n_pos"] / (s["n_pos"] + s["n_neg"])))
    if rule == "youden":
        return float(s["thr"][int(np.argmax(s["tp"] / s["n_pos"] - s["fp"] / s["n_neg"]))])
    n = s["n_pos"] + s["n_neg"]
    fn, tn = s["n_pos"] - s["tp"], s["n_neg"] - s["fp"]
    po = (s["tp"] + tn) / n
    pe = ((s["tp"] + s["fp"]) * s["n_pos"] + (fn + tn) * s["n_neg"]) / n ** 2
    with np.errstate(invalid="ignore", divide="ignore"):
        k = np.where(pe >= 1, -np.inf, (po - pe) / (1 - pe))
    return float(s["thr"][int(np.argmax(k))])


def cohen_kappa(a, b) -> float:
    """Chance-corrected agreement of two labellings of the same units, in either order."""
    a = np.asarray(a).astype(np.int64)
    b = np.asarray(b).astype(np.int64)
    n = len(a)
    both = int(np.sum((a == 1) & (b == 1)))
    neither = int(np.sum((a == 0) & (b == 0)))
    a_only = int(np.sum((a == 1) & (b == 0)))
    b_only = int(np.sum((a == 0) & (b == 1)))
    po = (both + neither) / n
    pe = ((both + a_only) * (both + b_only) + (b_only + neither) * (a_only + neither)) / n ** 2
    return float("nan") if pe >= 1 else float((po - pe) / (1 - pe))


def kappa_score(y, p, rule: str = "youden") -> float:
    """Cohen's kappa of a model's decisions against the observed response."""
    thr = decision_threshold(y, p, rule)
    if not np.isfinite(thr):
        return float("nan")
    return cohen_kappa(_labels(y), (np.asarray(p, dtype=np.float64) >= thr).astype(int))


def model_agreement(y, p_a, p_b, rule: str = "youden") -> dict:
    """Agreement between two models' decisions, with how often each is right where they differ."""
    y = _labels(y)
    ta, tb = decision_threshold(y, p_a, rule), decision_threshold(y, p_b, rule)
    if not (np.isfinite(ta) and np.isfinite(tb)):
        return dict(kappa=float("nan"), n=len(y), n_disagree=0, share_disagree=float("nan"),
                    a_right=0, b_right=0)
    da = (np.asarray(p_a, dtype=np.float64) >= ta).astype(int)
    db = (np.asarray(p_b, dtype=np.float64) >= tb).astype(int)
    diff = da != db
    return dict(kappa=cohen_kappa(da, db), n=len(y), n_disagree=int(diff.sum()),
                share_disagree=float(diff.mean()),
                a_right=int(np.sum(diff & (da == y))), b_right=int(np.sum(diff & (db == y))))
